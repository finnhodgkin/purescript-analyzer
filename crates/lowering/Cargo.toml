[package]
name = "lowering"
version = "0.1.0"
edition = "2021"

[dependencies]
id = { version = "0.1.0", path = "../id" }
indexing = { version = "0.1.0", path = "../indexing" }
indexmap = "2.7.1"
lexing = { version = "0.1.0", path = "../lexing" }
parsing = { version = "0.1.0", path = "../parsing" }
rowan = "0.16.1"
rustc-hash = "2.1.0"
syntax = { version = "0.1.0", path = "../syntax" }
