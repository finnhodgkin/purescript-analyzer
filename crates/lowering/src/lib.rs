use indexing::IndexingResult;
use syntax::cst;

mod algorithm;
mod concrete;
mod sourcemap;

pub use concrete::*;
pub use sourcemap::*;

pub fn lower(module: &cst::Module, index: &IndexingResult) {
    algorithm::lower_module(module, &index);
}
