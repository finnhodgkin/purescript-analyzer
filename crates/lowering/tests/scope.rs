mod shared;

use lowering::{ExpressionKind, LetBindingResolution, TermResolution, TypeKind};
use std::fmt::Write;
use test_each_file::test_each_file;

fn variable_scope_check(content: &str) -> String {
    let (_, ir, source, _) = shared::index_source(content);

    let mut snapshot = String::default();

    writeln!(snapshot, "Expressions:").unwrap();
    writeln!(snapshot).unwrap();
    for (expression_id, _) in ir.iter_expression() {
        let Some(ExpressionKind::Variable { resolution }) = ir.index_expression_kind(expression_id)
        else {
            continue;
        };
        let cst = &source[expression_id];
        let range = cst.syntax_node_ptr().text_range();
        writeln!(snapshot, "{:?}", range).unwrap();
        if let Some(resolution) = resolution {
            match resolution {
                TermResolution::Root(_) => {
                    writeln!(snapshot, "  resolves to top-level name").unwrap();
                }
                TermResolution::Binder(binder) => {
                    let cst = &source[*binder];
                    let range = cst.syntax_node_ptr().text_range();
                    writeln!(snapshot, "  resolved to binder {:?}", range).unwrap();
                }
                TermResolution::Let(LetBindingResolution { signature, equations }) => {
                    if let Some(signature) = signature {
                        let cst = &source[*signature];
                        let range = cst.syntax_node_ptr().text_range();
                        writeln!(snapshot, "  resolves to signature {:?}", range).unwrap();
                    }
                    for equation in equations.iter() {
                        let cst = &source[*equation];
                        let range = cst.syntax_node_ptr().text_range();
                        writeln!(snapshot, "  resolves to equation {:?}", range).unwrap();
                    }
                }
            }
        } else {
            writeln!(snapshot, "  resolves to nothing").unwrap();
        }
    }

    writeln!(snapshot, "\nTypes:\n").unwrap();

    for (type_id, _) in ir.iter_type() {
        let Some(TypeKind::Variable { resolution }) = ir.index_type_kind(type_id) else {
            continue;
        };
        let cst = &source[type_id];
        let range = cst.syntax_node_ptr().text_range();
        writeln!(snapshot, "{:?}", range).unwrap();
        if let Some(resolution) = resolution {
            let cst = &source[*resolution];
            let range = cst.syntax_node_ptr().text_range();
            writeln!(snapshot, "  resolves to {:?}", range).unwrap();
        } else {
            writeln!(snapshot, "  resolves to nothing").unwrap();
        }
    }

    snapshot
}

test_each_file! { in "./crates/lowering/tests/scope" as variable => |content: &str| {
    let snapshot = variable_scope_check(content);
    insta::assert_snapshot!(snapshot);
}}
