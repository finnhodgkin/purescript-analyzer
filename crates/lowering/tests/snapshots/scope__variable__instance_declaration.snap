---
source: crates/lowering/tests/scope.rs
expression: snapshot
snapshot_kind: text
---
Expressions:

77..87
  resolves to top-level name
169..181
  resolves to top-level name

Types:

138..140
  resolves to nothing
149..151
  resolves to nothing
